bootloader/grub.go

```diff
@@ -132,14 +138,169 @@ func (g *grub) SetBootVars(values map[string]string) error {
+	// or through an option in the bootloader
+	_, err := snapf.ReadFile("meta/force-kernel-extraction")
+	if g.uefiRunKernelExtraction || err == nil {
+		return extractKernelAssetsToBootDir(
+			g.extractedKernelDir(g.dir(), s),
+			s,
+			snapf,
+			assets,
+		)
 	}
 	return nil
 }
 
 func (g *grub) RemoveKernelAssets(s snap.PlaceInfo) error {
 	return removeKernelAssetsFromBootDir(g.dir(), s)
 }
+
+// ExtractedRunKernelImageBootloader helper methods
+
+func (g *grub) makeKernelEfiSymlink(s snap.PlaceInfo, name string) error {
+	// use a relative symlink destination so that it resolves properly, if grub
+	// is located at /run/mnt/ubuntu-boot or /boot/grub, etc.
+	extractedKernel := filepath.Join(
+		filepath.Base(s.MountFile()),
+		"kernel.efi",
+	)
+
+	// check that the kernel snap has been extracted already so we don't
+	// inadvertently create a dangling symlink
+	// expand the relative symlink from g.dir()
+	if !osutil.FileExists(filepath.Join(g.dir(), extractedKernel)) {
+		return fmt.Errorf(
+			"cannot enable %s at %s: %v",
+			name,
+			extractedKernel,
+			os.ErrNotExist,
+		)
+	}
+
+	return os.Symlink(
+		extractedKernel,
+		filepath.Join(g.dir(), name),
+	)
+}
+
+// unlinkKernelEfiSymlink will remove the specified symlink if it exists. Note
+// that if the symlink is "dangling", it will still remove the symlink without
+// returning an error. This is useful for example to disable a try-kernel that
+// was incorrectly created.
+func (g *grub) unlinkKernelEfiSymlink(name string) error {
+	symlink := filepath.Join(g.dir(), name)
+	// check that the symlink exists
+	if _, err := os.Lstat(symlink); err == nil {
+		// symlink exists, remove it unconditionally
+		return os.Remove(symlink)
+	}
+
+	// return more helpful error if the symlink doesn't exist
+	return fmt.Errorf("cannot disable kernel, symlink %s missing", symlink)
+}
+
+func (g *grub) readKernelSymlink(name string) (snap.PlaceInfo, error) {
+	// read the symlink from <grub-dir>/<name> to
+	// <grub-dir>/<snap-file-name>/<name> and parse the
+	// directory (which is supposed to be the name of the snap) into the snap
+	link := filepath.Join(g.dir(), name)
+
+	// check that the symlink is not dangling before continuing
+	if !osutil.FileExists(link) {
+		return nil, fmt.Errorf("cannot read dangling symlink %s", name)
+	}
+
+	targetKernelEfi, err := os.Readlink(link)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't read %s symlink: %v", link, err)
+	}
+
+	kernelSnapFileName := filepath.Base(filepath.Dir(targetKernelEfi))
+	sn, err := snap.ParsePlaceInfoFromSnapFileName(kernelSnapFileName)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"bad kernel snap file path at %q (from symlink %q), unable to parse into snap file name: %v",
+			kernelSnapFileName,
+			link,
+			err,
+		)
+	}
+	return sn, nil
+}
+
+// actual ExtractedRunKernelImageBootloader methods
+
+// EnableKernel will install a kernel.efi symlink on the bootloader pointing
```

Comment:
s/on the bootloader/in the bootloader partition/ ?

Reply:
tweaked the comment a little to make it clearer